src/gleam/time/duration.gleam

```diff
@@ -0,0 +1,124 @@
+import gleam/int
+import gleam/order
+import gleam/string
+
+// TODO: document
+pub opaque type Duration {
+  // When compiling to JavaScript ints have limited precision and size. This
+  // means that if we were to store the the timestamp in a single int the
+  // duration would not be able to represent very large or small durations.
+  // Durations are instead represented as a number of seconds and a number of
+  // nanoseconds.
+  //
+  // If you have manually adjusted the seconds and nanoseconds values the
+  // `normalise` function can be used to ensure the time is represented the
+  // intended way, with `nanoseconds` being positive and less than 1 second.
+  Duration(seconds: Int, nanoseconds: Int)
+}
+
+/// Ensure the duration is represented with `nanoseconds` being positive and
+/// less than 1 second.
+///
+/// This function does not change the amount of time that the duratoin refers
+/// to, it only adjusts the values used to represent the time.
+///
+fn normalise(duration: Duration) -> Duration {
+  let multiplier = 1_000_000_000
+  let nanoseconds = duration.nanoseconds % multiplier
+  let overflow = duration.nanoseconds - nanoseconds
+  let seconds = duration.seconds + overflow / multiplier
+  Duration(seconds, nanoseconds)
+}
+
+// TODO: docs
+pub fn compare(left: Duration, right: Duration) -> order.Order {
+  order.break_tie(
+    int.compare(left.seconds, right.seconds),
+    int.compare(left.nanoseconds, right.nanoseconds),
+  )
+}
+
+// TODO: docs
+pub fn difference(left: Duration, right: Duration) -> Duration {
+  Duration(right.seconds - left.seconds, right.nanoseconds - left.nanoseconds)
+  |> normalise
+}
+
+// TODO: docs
+pub fn add(left: Duration, right: Duration) -> Duration {
+  Duration(left.seconds + right.seconds, left.nanoseconds + right.nanoseconds)
+  |> normalise
+}
+
+// TODO: docs
+pub fn to_iso8601_string(duration: Duration) -> String {
+  let split = fn(total, limit) {
+    let amount = total % limit
+    let remainder = { total - amount } / limit
+    #(amount, remainder)
+  }
+  let #(seconds, rest) = split(duration.seconds, 60)
+  let #(minutes, rest) = split(rest, 60)
+  let #(hours, rest) = split(rest, 24)
+  let days = rest
+  let add = fn(out, value, unit) {
+    case value {
+      0 -> out
+      _ -> out <> int.to_string(value) <> unit
+    }
+  }
+  let output =
+    "P"
+    |> add(days, "D")
```

Comment:
I think including the day component for durations could cause some confusion since the length of a day can vary (e.g., 23h and 25h days are common during DST changes)

Reply:
Given that irregularly occurring longer minutes (and seconds for some systems) are also present can we use those also? My reasoning was that due to this duration being rendered without the present of a calendar they are all usable, while years are not as they are regularly of different lengths.

Reply:
I'm not a domain expert, and have never had to use this format, so not going to argue. Just some
thoughts:

I'm not sure what you mean by longer minutes/seconds. If you mean leap seconds then yes, that's the
same problem I'm worried about unless the library is going to completely ignore leap seconds. Other
than that, for me a duration is simply a number of seconds. Whether that number accurately
represents the amount of time it's meant to represent is irrelevant. Inaccurate system clocks, NTP
adjustments, or time dilation are not what I'm thinking about.

I think for me the main problem is that **if the same format is used for both durations and
periods**, then when I see "P1D" I can't know what it means without knowing which type it
represents. And if the hypothetical period type supports sub-day units then this is true even for
"PT1S": is it a duration of 1s or a period, which might actually correspond to a duration of 3601s
during a DST transition?

Reply:
Yes that's nature of iso8601, the value it represents cannot be determined without additional context.